influxdata/tokenizer.go

```diff
@@ -47,20 +69,389 @@ type Tokenizer struct {
+		t.take(fieldSeparatorSpace)
+		t.section = FieldSection
+		return nil, nil, nil
+	}
+	tagKey := t.takeEsc(tagKeyChars, &tagKeyEscapes.revTable)
+	if len(tagKey) == 0 {
+		return nil, nil, t.syntaxErrorf("empty tag name")
+	}
+	if !t.ensure(1) || t.at(0) != '=' {
+		return nil, nil, t.syntaxErrorf("expected '=' after tag key %q, but got %q instead", tagKey, t.at(0))
+	}
+	t.advance(1)
+	tagVal := t.takeEsc(tagValChars, &tagValEscapes.revTable)
+	if len(tagVal) == 0 {
+		return nil, nil, t.syntaxErrorf("expected value after tag name")
+	}
+	if !t.ensure(1) {
+		return nil, nil, t.syntaxErrorf("expected value after tag name")
+	}
+	if err := t.advanceTagComma(); err != nil {
+		return nil, nil, err
+	}
+	return tagKey, tagVal, nil
+}
+
+// advanceTagComma consumes a comma after a measurement
+// or a tag value, making sure it's not followed by whitespace.
+func (t *Tokenizer) advanceTagComma() error {
+	if !t.ensure(1) {
+		return nil
+	}
+	nextc := t.at(0)
+	if nextc != ',' {
+		return nil
+	}
+	// If there's a comma, there's a tag, so check that there's the start
+	// of a tag name there.
+	t.advance(1)
+	if !t.ensure(1) {
+		return t.syntaxErrorf("expected tag name after comma; got end of input")
+	}
+	if whitespace.get(t.at(0)) {
+		return t.syntaxErrorf("expected tag name after comma; got white space instead")
+	}
+	return nil
+}
+
+// NextFieldBytes returns the next field in the entry.
+// If there are no more fields, it returns all zero values.
+// Note that this must be called before Time because
+// fields precede the timestamp in the line-protocol entry.
+//
+// The returned value slice may not be fully checked: to
+// check its validity, use NewValueFromBytes, or use NextField.
+func (t *Tokenizer) NextFieldBytes() (key []byte, kind ValueKind, value []byte, err error) {
+	if ok, err := t.advanceToSection(FieldSection); err != nil {
+		return nil, Unknown, nil, err
+	} else if !ok {
+		return nil, Unknown, nil, nil
+	}
+	fieldKey := t.takeEsc(fieldKeyChars, &fieldKeyEscapes.revTable)
+	if len(fieldKey) == 0 {
+		return nil, Unknown, nil, t.syntaxErrorf("expected field key but none found")
+	}
+	if !t.ensure(1) {
+		return nil, Unknown, nil, t.syntaxErrorf("want '=' after field name, found end of input ")
+	}
+	if nextc := t.at(0); nextc != '=' {
+		return nil, Unknown, nil, t.syntaxErrorf("want '=' after field name, found %q", nextc)
+	}
+	t.advance(1)
+	if !t.ensure(1) {
+		return nil, Unknown, nil, t.syntaxErrorf("expected field value, found end of input")
+	}
+	var fieldVal []byte
+	var fieldKind ValueKind
+	switch t.at(0) {
+	case '"':
+		// Skip leading quote.
+		t.advance(1)
+		fieldVal = t.takeEsc(fieldStringValChars, &fieldStringValEscapes.revTable)
+		fieldKind = String
+		if !t.ensure(1) {
+			return nil, Unknown, nil, t.syntaxErrorf("expected closing quote for string field value, found end of input")
+		}
+		if t.at(0) != '"' {
+			// This shouldn't be able to happen, as all characters are allowed in a string.
```

Comment:
I don't believe single quotes are allowed nor unescaped backslashes. Dunno what your corpus says.

Reply:
Single quotes are allowed in strings (the doc is misleading here) - you just can't use single quotes to quote a string.
You can have unescaped backslashes - an backslash followed by an unknown escape character represents exactly a backslash followed by that character. So `\q` encodes to just itself. Not that `fieldStringValChars` is _exactly_ everything except `"`.